Add GEDCOM 7 Extensions Support for GRAMPS Compatibility

glamberson opened the pull request:
# Add GEDCOM 7 Extensions Support for GRAMPS Compatibility

## Summary

This PR adds support for three GEDCOM 7 extensions that enable full compatibility with GRAMPS' advanced data model, addressing the long-standing issue of data loss when exchanging GEDCOM files with GRAMPS.

## Motivation

GRAMPS has a richer data model than standard GEDCOM supports, leading to data loss during import/export. This PR implements support for GEDCOM 7 extensions that preserve:
- Shared events (multiple people at one census/burial/etc.)
- Evidence containers (separation of sources from conclusions)
- Template-based citations (following Evidence Explained standards)
- Rich text formatting in notes

## Changes

### New Modules
- `occurrence.py` - Handles shared event records (`_OCUR`/`_OCREF`)
- `evidence.py` - Handles evidence containers (`_EVID`) 
- `citation_templates.py` - Handles citation templates (`_TMPLT`)
- `process_enhanced.py` - Enhanced processor with extension detection
- `individual_enhanced.py` - Enhanced individual handler
- `note_enhanced.py` - HTML subset to StyledText conversion

### Key Features
1. **Extension Detection**: Reads SCHMA declarations to identify registered extensions
2. **Shared Events**: Maps `_OCUR` records to GRAMPS Event objects
3. **Evidence Management**: Maps `_EVID` to GRAMPS research notes
4. **Citation Templates**: Preserves template structure in source attributes
5. **HTML Formatting**: Converts GEDCOM 7 HTML subset to GRAMPS StyledText

### Testing
- Comprehensive test suite in `test/test_extensions.py`
- Example GEDCOM files demonstrating each extension
- Tests for extension interoperability

## Example

```gedcom
# Shared census event
0 @O1@ _OCUR
1 TYPE Census
1 DATE 1850
1 _PART @I1@
2 ROLE Head

# Person references the event
0 @I1@ INDI
1 NAME John /Smith/
1 _OCREF @O1@
```

## Compatibility

- Fully backward compatible - unknown extensions are ignored
- Follows GEDCOM 7 specification for extension handling
- Maintains existing behavior for standard GEDCOM files

## Related Issues

- Addresses GRAMPS Bug [#12226](https://gramps-project.org/bugs/view.php?id=12226)
- Implements extensions from [FamilySearch/GEDCOM#663](https://github.com/FamilySearch/GEDCOM/issues/663)
- Supports extensions registered in [PR #173](https://github.com/FamilySearch/GEDCOM-registries/pull/173)

## Documentation

- See `GEDCOM7_EXTENSIONS_SUPPORT.md` for detailed documentation
- Extension specifications:
  - [gedcom-occurrences](https://github.com/glamberson/gedcom-occurrences)
  - [gedcom-evidence](https://github.com/glamberson/gedcom-evidence)
  - [gedcom-citations](https://github.com/dthaler/gedcom-citations)

## Future Work

- Add export support (currently import-only)
- Support additional extensions as they become available
- Add user configuration for extension handling

## Testing Instructions

1. Install dependencies: `pip install -r requirements-dev.txt`
2. Run tests: `pytest test/test_extensions.py`
3. Test with example files in `test/data/`

This PR represents a significant step forward in GEDCOM 7 adoption by demonstrating how extensions can solve real compatibility issues between genealogy applications.

DavidMStraub commented:
Sorry, but I am bit shocked by this huge PR which comes with any prior warning or any issue where implementation details could have been discussed, and which comes with huge markdown files with far-too-perfect formatting and emoji support that make it very obvious that a big part of this PR was prepared with the help of AI. Also, as is typical for vibe coding tools, it adds Markdown files all over the place, although it should be obvious to anyone who has cared to look at the contents of this repository that this is not the way things are documented here.

So, before even starting to review this, I want you to be fully transparent here about which tools were used and what the prompts were, please.

Thank you.

glamberson commented:
Hi David, LEt's see. I've been working on this for more than 15 years, and everyone who knows me knows I'm a systems and networking guy, not a coder. I make no bones about using AI tools as I can build things I previously had to wait aorund for developers to do or do it myself which I hate. The discussion you'll see is in the bug report I submitted and this very work is responsive to. that I submitted YEARS ago. So there's your discussion. 

I'm sorry you have an anti-AI thing, and that's fine. It doesn't match your format or your usual way of doing things? Sorry.  That's not very relevant to the work. If there's something you don't like, let me know and I'll be glad to do it. I'm a wonderful writer. I can write a book in aobut ten minutes.  However, in the meantime I hope you'll get over your piquancy and pay attention to what is in front of you. 

As for the prompts I used, get real. I've been doing this for 40 years or so so I know how to explain what I ant. I was the very brains behind BetterGEDCOM and directly pushed to get anything done in this arena when nothing was for decades. So I"m not doing this like a toddler.

glamberson commented:
That sounds great to me. Yes, it's a lot, but I"m between projects, and I'm in a hurry to get this done on my end before I move to the next thing. If it takes time and interation that's perfectly ok. But I would've done the entire thing myself had I not seen you already started a gedcom7 addon.

Please tell me how you'd like to proceed, and I'll be happy to accommodate.

DavidMStraub commented:
Added [code of conduct](https://github.com/DavidMStraub/gramps-gedcom7/blob/main/CODE_OF_CONDUCT.md) as for my other repos, please do hold me accountable to it as well.

> Please tell me how you'd like to proceed, and I'll be happy to accommodate.
 
I propose you give me a couple of days to digest and I will come back with a (better informed) suggestion.

glamberson commented:
Absolutely. Take your time. Please don't take any further submissions by me in the future as being of urgency or too much. It's just that when I have time to focus on something I have to get done what I can. Thanks.

glamberson commented:
I've pushed fixes for all 10 type errors. The changes include:

- Import fixes (handle_place, EventRoleType)
- Added registered_extensions field to ImportSettings  
- Fixed tuple handling in several places
- Replaced missing handle_fam_link with inline logic
- Corrected function signatures and return types

All type errors should now be resolved. The code passes mypy type checking locally.

DavidMStraub commented:
Thanks for letting me know! I would be happy to discuss and learn about this more in the future, so far extensions is something I have pretty much ignored.

My priority for the next couple of weeks will be to finish the implementation of the standard tags, so at least we have a first usable (or rather testable) version of an import addon.
